spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,296 @@
+package org.apache.spark.sql.cassandra
+
+import org.apache.spark.Logging
+
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.cassandra.CassandraSQLRow.CassandraSQLRowReader
+import org.apache.spark.sql.sources._
+import org.apache.spark.sql.types._
+import org.apache.spark.sql.{DataFrame, Row, SQLContext}
+
+import com.datastax.spark.connector._
+import com.datastax.spark.connector.cql.{CassandraConnector, ColumnDef}
+import com.datastax.spark.connector.rdd.ReadConf
+import com.datastax.spark.connector.writer.{WriteConf, SqlRowWriter}
+
+import DataTypeConverter._
+
+/**
+ *  Implements [[BaseRelation]] and [[InsertableRelation]].
+ *  It inserts data into Cassandra table. It also provides some
+ *  helper methods and variables for sub class to use.
+ */
+private[cassandra] abstract class BaseRelationImpl(
+    tableRef: TableRef,
+    connector: CassandraConnector,
+    readConf: ReadConf,
+    writeConf: WriteConf,
+    userSpecifiedSchema: Option[StructType],
+    override val sqlContext: SQLContext)
+  extends BaseRelation
+  with InsertableRelation
+  with Serializable
+  with Logging {
+
+  protected[this] val tableDef =
+    sqlContext.getCassandraSchema(tableRef.cluster.getOrElse("default"))
+      .keyspaceByName(tableRef.keyspace).tableByName(tableRef.table)
+
+  override def schema: StructType = {
+    userSpecifiedSchema.getOrElse(StructType(tableDef.allColumns.map(toStructField)))
+  }
+
+  implicit val cassandraConnector = connector
+  implicit val readconf = readConf
+  protected[this] val baseRdd =
+    sqlContext.sparkContext.cassandraTable[CassandraSQLRow](tableRef.keyspace, tableRef.table)
+
+  override def insert(data: DataFrame, overwrite: Boolean): Unit = {
+    if (overwrite) {
```

Comment:
I think the truncate is the right thing to do after checking the JDBC implementation in spark 1.4 which calls a truncate

https://github.com/apache/spark/blob/213a6f30fee4a1c416ea76b678c71877fd36ef18/sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala#L1750-L1763